File: beginner0107/BFS/숫자변환하기.java
import java.util.LinkedList;
import java.util.Queue;

public class 숫자변환하기 {
    public static int bfs(int start, int end, int increment) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        int[] visited = new int[end + 1];

        while (!queue.isEmpty()) {
            int node = queue.poll();
            int oneStep = node + increment;
            int twoStep = node * 2;
            int threeStep = node * 3;

            for (int i : new int[]{oneStep, twoStep, threeStep}) {
                if (i <= end && visited[i] == 0) {
                    visited[i] = visited[node] + 1;
                    queue.add(i);
                }
            }
        }
        return visited[end];
    }

    public static int solution(int x, int y, int n) {
        if (x == y) {
            return 0;
        }
        int answer = bfs(x, y, n);
        if (answer == 0) {
            return -1;
        }
        return answer;
    }
}

File: beginner0107/BFS/숫자변환하기.py
from collections import deque

def bfs(start, end, increment):
    queue = deque([start])
    visited = [0 for _ in range(end + 1)]

    while queue:
        node = queue.popleft()
        one_step = node + increment
        two_step = node * 2
        three_step = node * 3

        for i in [one_step, two_step, three_step]:
            if i <= end and visited[i] == 0:
                visited[i] = visited[node] + 1
                queue.append(i)
    return visited[end]


def solution(x, y, n):
    if x == y: 
        return 0
    answer = bfs(x, y, n)
    if answer == 0:
        return -1
    return answer